### PR TITLE
Experimental conda packaging (work in progress)

### PR DESCRIPTION
**Pull request by mara004**

### CODE REVIEW

**Comment by bblanchon** [comment]
Thanks!
That's a good starting point.
However, I think I'll do the conda build in a dedicated job instead of adding it to the build script.
I'll give it a try and get back to you.

**Comment by mara004** [comment]
> However, I think I'll do the conda build in a dedicated job instead of adding it to the build script.

Sure, good point. I just put it into the stage step for simplicity for now (this is only a draft after all).

File: steps/07-stage.sh [hunk 1]
```diff
@@ -19,33 +21,33 @@ sed "s/#VERSION#/${VERSION:-0.0.0.0}/" <"$PATCHES/PDFiumConfig.cmake" >"$STAGING
...
-cp -R "$SOURCE/public" "$STAGING/include"
+cp -R "$SOURCE/public" -T "$STAGING/include"
 rm -f "$STAGING/include/DEPS"
 rm -f "$STAGING/include/README"
 rm -f "$STAGING/include/PRESUBMIT.py"
 
 case "$OS" in
   android|linux)
-    mv "$BUILD/libpdfium.so" "$STAGING_LIB"
+    cp "$BUILD/libpdfium.so" "$STAGING_LIB"
```
Comment (mara004) [inline steps/07-stage.sh:31]: note: the `mv` -> `cp` changes are unrelated; I just did this for local testing so a single step can be run multiple times.

File: steps/07-stage.sh [hunk 2]
```diff
@@ -19,33 +21,33 @@ sed "s/#VERSION#/${VERSION:-0.0.0.0}/" <"$PATCHES/PDFiumConfig.cmake" >"$STAGING
 
 cp "$SOURCE/LICENSES" "$STAGING/LICENSE"
 cp "$BUILD/args.gn" "$STAGING"
-cp -R "$SOURCE/public" "$STAGING/include"
+cp -R "$SOURCE/public" -T "$STAGING/include"
```
Comment (mara004) [inline steps/07-stage.sh:24]: Without `-T`, the `public/` directory would be moved into `include/` if it already exists.
While this is irrelevant under CI circumstances, it's a bit confusing for local testing.

**Comment by bblanchon** [comment]
I tried to make sense of this, but I need clarification on all the folder manipulation.
How can I build the one package that includes all the binaries, given that all the required files are on disk?
It should just be a call to `conda build`, nothing else, right?

**Comment by mara004** [comment]
>  I tried to make sense of this, but I need clarification on all the folder manipulation.
> How can I build the one package that includes all the binaries, given that all the required files are on disk?
> It should just be a call to conda build, nothing else, right?

I'm not sure what you mean?

With conda we build one package for each platform, not a single big package for all platforms.
For packages that match the host platform it will be just a call to `conda build`; for non-host packages we have to create an intermediary fake host package, then find and convert it using `conda convert`.

**Comment by mara004** [comment]
Also see my edited post above regarding what I believe you intend to do.
I think a key point is the source path. If you want to build on the root runner, it will have to be made dynamic using an env var.

File: steps/07-stage.sh [hunk 3]
```diff
@@ -19,41 +21,103 @@ sed "s/#VERSION#/${VERSION:-0.0.0.0}/" <"$PATCHES/PDFiumConfig.cmake" >"$STAGING
...
+      ;;
+  esac
+  
+  if [ "$CONDA_TARGET" == "none" ]; then
+    mkdir -p conda/staging/$CONDA_HOST/
+    cp conda/out/$CONDA_HOST/*.tar.bz2 conda/staging/$CONDA_HOST/
+  else
+    conda convert conda/out/$CONDA_HOST/*.tar.bz2 -p $CONDA_TARGET -o conda/out/
+    mkdir -p conda/staging/$CONDA_TARGET/
+    cp conda/out/$CONDA_TARGET/*.tar.bz2 conda/staging/$CONDA_TARGET/
```
Comment (mara004) [inline steps/07-stage.sh:118]: (note for clarity: the `*` will only find a single file - my original intent here was to build each conda package in the individual build jobs, put it into `conda/staging/` and transfer to the root runner using `upload-artifact`)

**Comment by bblanchon** [comment]
I made some progress, as you can see on this build:
https://github.com/bblanchon/pdfium-binaries/actions/runs/5972658236/job/16203843717

I'm facing issues with the Windows build, and I wonder if creating a Windows package from a Linux host is really supported.

If I copy `pdfium.dll` in `bin/` here is what I get:

```
+ conda convert -f -p win-32 out/linux-64/pdfium-binaries-118.0.5961-0.tar.bz2 -o staging/
Converting pdfium-binaries-118.0.5961-0.tar.bz2 from linux-64 to win-32

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/opt/conda/lib/python3.11/site-packages/conda/exception_handler.py", line 17, in __call__
        return func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda/cli/main.py", line 64, in main_subshell
        exit_code = do_call(args, parser)
                    ^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda/cli/conda_argparse.py", line 143, in do_call
        result = plugin_subcommand.action(getattr(args, "_args", args))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/plugin.py", line 16, in convert
        execute(*args, **kwargs)
      File "/opt/conda/lib/python3.11/site-packages/conda_build/cli/main_convert.py", line 129, in execute
        api.convert(f, **args.__dict__)
      File "/opt/conda/lib/python3.11/site-packages/conda_build/api.py", line 404, in convert
        return conda_convert(
               ^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 874, in conda_convert
        convert_from_unix_to_windows(
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 732, in convert_from_unix_to_windows
        update_paths_file(temp_dir, target_platform="win")
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 415, in update_paths_file
        path["sha256"] = update_executable_sha(temp_dir, path["_path"])
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 349, in update_executable_sha
        with open(os.path.join(package_directory, executable_path), "rb") as script_file:
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpv93itaar/Scripts/pdfium-script.py'
```

If I copy `pdfium.dll` in `lib/` here is what I get:

```
+ conda convert -f -p win-32 out/linux-64/pdfium-binaries-118.0.5961-0.tar.bz2 -o staging/
Converting pdfium-binaries-118.0.5961-0.tar.bz2 from linux-64 to win-32

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/opt/conda/lib/python3.11/site-packages/conda/exception_handler.py", line 17, in __call__
        return func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda/cli/main.py", line 64, in main_subshell
        exit_code = do_call(args, parser)
                    ^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda/cli/conda_argparse.py", line 143, in do_call
        result = plugin_subcommand.action(getattr(args, "_args", args))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/plugin.py", line 16, in convert
        execute(*args, **kwargs)
      File "/opt/conda/lib/python3.11/site-packages/conda_build/cli/main_convert.py", line 129, in execute
        api.convert(f, **args.__dict__)
      File "/opt/conda/lib/python3.11/site-packages/conda_build/api.py", line 404, in convert
        return conda_convert(
               ^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 874, in conda_convert
        convert_from_unix_to_windows(
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 732, in convert_from_unix_to_windows
        update_paths_file(temp_dir, target_platform="win")
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 409, in update_paths_file
        path["_path"] = update_lib_path(path["_path"], "win")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 243, in update_lib_path
        python_version = retrieve_python_version(path)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/conda/lib/python3.11/site-packages/conda_build/convert.py", line 108, in retrieve_python_version
        with open(path_file) as index_file:
             ^^^^^^^^^^^^^^^
    FileNotFoundError: [Errno 2] No such file or directory: 'lib/pdfium.dll/info/index.json'
```

Anyway, I spent way too much time on this, and I need to get back to work for the moment.

**Comment by bblanchon** [comment]
I am tired of wasting time on these awful tools (`conda build` and `conda convert`).
What a mess to create a tarball!

I would probably be better off building the tarball from a script since [the specification is available](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html).
What do you think?

**Comment by mara004** [comment]
> Anyway, I spent way too much time on this, and I need to get back to work for the moment.

> I am tired of wasting time on these awful tools (conda build and conda convert).
> What a mess to create a tarball!

I'm sorry for the inconveniences; I too have been feeling this way. It is a mess indeed, these tools are slow and not well designed. conda packaging just seems to be very centered towards the native hosts model of anaconda's CI...
Nevertheless, I would like to finally get this through because I've put in a lot of effort already.

> I would probably be better off building the tarball from a script since [the specification is available](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html).
What do you think?

I've once thought about this too, but frustration aside, I don't think it's reasonable. Probably this would be way more work and sufficiently difficult to get right.

> I'm facing issues with the Windows build, and I wonder if creating a Windows package from a Linux host is really supported.

If it doesn't work it probably isn't supported, unfortunately.
How about just building the windows packages natively on windows and use the `bld.bat` as shown in this PR?
Maybe integrating this into the build steps as I initially thought was better after all?

(FWIW converting from linux to windows did work with python though, see [pypdfium2's conda bundling experiment](https://github.com/pypdfium2-team/pypdfium2/compare/main...conda))

BTW I like the `-o staging/`, this avoids some of my strange directory logic.

**Comment by mara004** [comment]
Ideally the recipe should just provide an option to set the platform directly, but I couldn't find anything, so probably this is not supported.

**Comment by bblanchon** [comment]
> How about just building the windows packages natively on windows

I prefer not to, because:

1. I don't want conda to be required to build other packages
2. I don't want the build to fail if there is a problem with conda
3. I want a dedicated job to upload to anaconda, so it makes sense to build the packages there
4. I want to keep each task isolated (imagine if we did the same with all packages formats)
5. I'm using the same pattern for NuGet and it's perfect

BTW, I also tied the undocumented command `conda package`, which looked promising, but I could not understand how it's supposed to work.

**Comment by mara004** [comment]
OK, I certainly agree with that rationale, it should not be part of the main build.
However, we can still divide the isolated receiver job into native windows/linux/mac hosts, no?
Not very nice, I agree, but should be possible?

**Comment by bblanchon** [comment]
> Not very nice, I agree, but should be technically possible?

Sure, but is this better than writing our own packaging script, I'm not sure.

BTW, after we have the packages, how are we supposed to upload them?
Please don't tell me a Linux machine cannot upload Windows packages 🤪

**Comment by mara004** [comment]
(BTW, sorry for the trouble. Disclaimer: I neither use nor like conda. Just want to do it in favor of the users who are requesting it.)

**Comment by mara004** [comment]
> Sure, but is this better than writing our own packaging script, I'm not sure.

Perhaps you're right, the `conda2` branch looks promising!
